File: src/1192.go
package main

import "fmt"

func main() {
	connections := [][]int{{0, 1}, {1, 2}, {2, 0}, {1, 3}}
	fmt.Println(criticalConnections(4, connections))
}

func criticalConnections(n int, connections [][]int) [][]int {
	adjList := make([][]int, n)
	for _, edge := range connections {
		adjList[edge[0]] = append(adjList[edge[0]], edge[1])
		adjList[edge[1]] = append(adjList[edge[1]], edge[0])
	}
	timer := 1
	bridges := make([][]int, 0)

	tin := make([]int, n)
	fup := make([]int, n)
	var dfs func(int, int)

	dfs = func(u int, p int) {
		tin[u], fup[u] = timer, timer
		timer++

		for _, v := range adjList[u] {
			if v == p {
				continue
			}
			if tin[v] != 0 {
				fup[u] = min(fup[u], tin[v])
			} else {
				dfs(v, u)
				fup[u] = min(fup[u], fup[v])
				if tin[u] < fup[v] {
					bridges = append(bridges, []int{u, v})
				}
			}
		}
	}

	dfs(0, -1)
	return bridges
}

File: 1192.cpp
class Solution {
public:

    vector<vector<int>> g;
    vector<bool> visited;
    vector<int> time;
    vector<int> low;
    vector<vector<int>> bridges;
    int timer = 0;

    void dfs(const int u, const int parent = -1) {
        if (visited[u]) {
            return;
        }
        visited[u] = true;
        time[u] = low[u] = timer++;
        for (int v : g[u]) {
            if (v == parent) {
                continue;
            }
            if (visited[v]) {
                low[u] = min(low[u], time[v]);
            }
            else {
                dfs(v, u);
                low[u] = min(low[u], low[v]);
                if (low[v] > time[u]) {
                    bridges.push_back({ u, v });
                }
            }
        }
    }
    vector<vector<int>> criticalConnections(const int n, const vector<vector<int>>& connections) {
        g.resize(n);
        for (const auto& e : connections) {
            g[e[0]].push_back(e[1]);
            g[e[1]].push_back(e[0]);
        }
        visited.assign(n, false);
        time.assign(n, -1);
        low.assign(n, -1);
        for (int i = 0; i < n; ++i) {
            if (!visited[i]) {
                dfs(i);
            }
        }
        return bridges;
    }
};